core/cog-gamepad-manette.c

```diff
@@ -0,0 +1,319 @@
+/*
+ * gamepad.h
+ * Copyright (C) 2022 Igalia S.L.
+ *
+ * Distributed under terms of the MIT license.
+ */
+
+#include "cog-gamepad.h"
+
+#include <libmanette.h>
+#include <linux/input-event-codes.h>
+#include <wpe/wpe.h>
+
+typedef struct _Provider Provider;
+typedef struct _Gamepad  Gamepad;
+
+struct _Provider {
+    struct wpe_gamepad_provider *wk_provider;
+    ManetteMonitor              *monitor;
+};
+
+struct _Gamepad {
+    struct wpe_gamepad *wk_gamepad;
+    ManetteDevice      *device;
+};
+
+static void *
+provider_create(struct wpe_gamepad_provider *wk_provider)
+{
+    Provider *provider = g_new0(Provider, 1);
+    provider->wk_provider = wk_provider;
+
+    return provider;
+}
+
+static void
+provider_destroy(void *data)
+{
+    Provider *provider = data;
+
+    if (!provider)
+        return;
+
+    g_clear_object(&provider->monitor);
+    g_free(provider);
+}
+
+static void
+add_device(Provider *provider, ManetteDevice *device)
+{
+    g_debug("gamepad: %s - connected", manette_device_get_name(device));
+
+    wpe_gamepad_provider_dispatch_gamepad_connected(provider->wk_provider, GPOINTER_TO_UINT(device));
+}
+
+static gboolean
+add_connected_devices(gpointer data)
+{
+    Provider                     *provider = data;
+    g_autoptr(ManetteMonitorIter) iter = manette_monitor_iterate(provider->monitor);
+    ManetteDevice                *device;
+
+    while (manette_monitor_iter_next(iter, &device)) {
+        add_device(provider, device);
+    };
```

Comment:
Please remove this stray semicolon before merging. The braces could be removed because there is only one statement, but I have no strong preference in this regard.